File: lib/crowding/features.ts
/**
 * Time and stop features for bus fullness heuristic.
 * Madison timezone (America/Chicago) for rush/dismissal.
 */

import { DISMISSAL_WINDOWS } from "./dismissal-windows"
import { getStopDensity, type StopDensityLevel } from "./stop-density"

// Rush hour windows (local 24h): morning 7–9, evening 17–18
const RUSH_HOURS = [
  [7, 9],
  [17, 18],
]

export type TimeFeatures = {
  hour: number
  minute: number
  day_of_week: number // 0 = Sunday, 5 = Saturday
  is_weekend: boolean
  in_rush_hour: boolean
  in_dismissal_window: boolean
}

export function getTimeFeatures(now: Date = new Date()): TimeFeatures {
  // Madison is America/Chicago
  const formatter = new Intl.DateTimeFormat("en-CA", {
    timeZone: "America/Chicago",
    hour: "numeric",
    minute: "numeric",
    hour12: false,
    weekday: "short",
  })
  const parts = formatter.formatToParts(now)
  const hour = parseInt(parts.find((p) => p.type === "hour")?.value ?? "0", 10)
  const minute = parseInt(parts.find((p) => p.type === "minute")?.value ?? "0", 10)
  const weekday = parts.find((p) => p.type === "weekday")?.value ?? "Mon"
  const dayMap: Record<string, number> = {
    Sun: 0,
    Mon: 1,
    Tue: 2,
    Wed: 3,
    Thu: 4,
    Fri: 5,
    Sat: 6,
  }
  const day_of_week = dayMap[weekday] ?? 1
  const is_weekend = day_of_week === 0 || day_of_week === 6

  const in_rush_hour = RUSH_HOURS.some(([start, end]) => hour >= start && hour < end)

  const timeStr = `${String(hour).padStart(2, "0")}:${String(minute).padStart(2, "0")}`
  const in_dismissal_window =
    !is_weekend &&
    DISMISSAL_WINDOWS.some((w) => timeStr >= w.start && timeStr <= w.end)

  return {
    hour,
    minute,
    day_of_week,
    is_weekend,
    in_rush_hour,
    in_dismissal_window,
  }
}

export function getStopDensityFeature(stopNameOrId: string): StopDensityLevel {
  return getStopDensity(stopNameOrId)
}

File: backend/crowding/features.py
"""
Time and stop features for bus fullness heuristic.
Madison timezone (America/Chicago) for rush/dismissal.
"""

from __future__ import annotations

from datetime import datetime
from zoneinfo import ZoneInfo

from .dismissal_windows import DISMISSAL_WINDOWS
from .stop_density import get_stop_density, StopDensityLevel

MADISON_TZ = ZoneInfo("America/Chicago")

# Rush: morning 7–9, evening 17–18 (local 24h)
RUSH_HOURS = [(7, 9), (17, 18)]


def get_time_features(now: datetime | None = None) -> dict:
    if now is None:
        now = datetime.now(MADISON_TZ)
    else:
        now = now.astimezone(MADISON_TZ)

    hour = now.hour
    minute = now.minute
    day_of_week = now.weekday()  # 0 Mon .. 6 Sun
    is_weekend = day_of_week >= 5  # Sat=5, Sun=6

    in_rush_hour = any(start <= hour < end for start, end in RUSH_HOURS)

    time_str = f"{hour:02d}:{minute:02d}"
    in_dismissal_window = (
        not is_weekend
        and any(w["start"] <= time_str <= w["end"] for w in DISMISSAL_WINDOWS)
    )

    return {
        "hour": hour,
        "minute": minute,
        "day_of_week": day_of_week,
        "is_weekend": is_weekend,
        "in_rush_hour": in_rush_hour,
        "in_dismissal_window": in_dismissal_window,
    }


def get_stop_density_feature(stop_name_or_id: str) -> StopDensityLevel:
    return get_stop_density(stop_name_or_id)
